telemetry/log.go

```diff
@@ -0,0 +1,42 @@
+package telemetry
+
+import (
+	"context"
+)
+
+type ctxKey int
+
+const logCtxKey ctxKey = iota
+
+// Field is a structured logging field..
+type Field struct {
+	Name  string
+	Value any
+}
+
+// Log is a interface for structure logging.
+type Log interface {
+	Debug(ctx context.Context, message string, fields ...Field)
+	Info(ctx context.Context, message string, fields ...Field)
+	Error(ctx context.Context, message string, err error, fields ...Field)
+}
+
+// ContextWithLog returns a new context that carries the supplied Log.
+func ContextWithLog(ctx context.Context, l Log) context.Context {
+	return context.WithValue(ctx, logCtxKey, l)
+}
+
+// LogFromContext returns the Log value stored in ctx, if any.
+// If a value is not found, a nop Log is returned.
+func LogFromContext(ctx context.Context) Log {
+	if log, ok := ctx.Value(logCtxKey).(Log); ok {
+		return log
+	}
+	return nopLog{}
```

Comment:
Will every context have a logger for the most part? Or do we have an option to set the default logger to be returned?

Reply:
As long as these steps are taken:
1. https://github.com/zenledger-io/zazen/blob/d4b7c58a1d99c1de874bc1165e52bbae1cd9412e/examples/api/main.go#L44
2. https://github.com/zenledger-io/zazen/blob/d4b7c58a1d99c1de874bc1165e52bbae1cd9412e/examples/api/main.go#L73-L75

Everything inside the service will have a logger.